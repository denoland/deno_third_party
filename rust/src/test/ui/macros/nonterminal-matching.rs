// Copyright 2018 The Rust Project Developers. See the COPYRIGHT
// file at the top-level directory of this distribution and at
// http://rust-lang.org/COPYRIGHT.
//
// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
// option. This file may not be copied, modified, or distributed
// except according to those terms.

// Check that we are refusing to match on complex nonterminals for which tokens are
// unavailable and we'd have to go through AST comparisons.

#![feature(decl_macro)]

macro simple_nonterminal($nt_ident: ident, $nt_lifetime: lifetime, $nt_tt: tt) {
    macro n(a $nt_ident b $nt_lifetime c $nt_tt d) {
        struct S;
    }

    n!(a $nt_ident b $nt_lifetime c $nt_tt d);
}

macro complex_nonterminal($nt_item: item) {
    macro n(a $nt_item b) {
        struct S;
    }

    n!(a $nt_item b); //~ ERROR no rules expected the token `enum E { }`
}

simple_nonterminal!(a, 'a, (x, y, z)); // OK

complex_nonterminal!(enum E {});

fn main() {}
