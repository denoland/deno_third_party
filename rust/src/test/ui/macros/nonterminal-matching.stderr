error: no rules expected the token `enum E { }`
  --> $DIR/nonterminal-matching.rs:19:10
   |
LL |     n!(a $nt_item b); //~ ERROR no rules expected the token `enum E { }`
   |          ^^^^^^^^ no rules expected this token in macro call
...
LL | complex_nonterminal!(enum E {});
   | -------------------------------- in this macro invocation

error: aborting due to previous error

